pkg/controller/node_availability.go

```diff
@@ -43,6 +44,8 @@ func (r *ReconcileNodeAvailability) Reconcile(ctx context.Context, request recon
 		return reconcile.Result{Requeue: true}, nil
 	}
 
+	ctx, workflow := azureutils.NewWorkflow(ctx, consts.NodeNameLabel, request.Name)
```

Comment:
Missing defer? But do we really need to create a workflow here if we create on in `tryCreateFailedReplicas`?

Reply:
I was treating the node availability as the workflow initiator to trigger the "recreate failed replicas` workflow. In such case, the workflow does not finish at the end of `Reconcile` function.

Reply:
> I was treating the node availability as the workflow initiator to trigger the "recreate failed replicas `workflow`.

Right now, the node availability controller has only one task, but in the future may have more. In general, we should feel free to view a reconciler as the workflow initiator when has it has one purpose. However, this reconciler starts an asynchronous to its execution and I view it as more akin to an event trigger than an initiator of activitiy. Also, we are unconditionally creating a new workflow when we only need to if we actually call `triggerCreateFailedReplicas`.

In the future, we may also want to trigger replica recreation from other controllers on different events than schedulability. In that case, starting the workflow in `triggerCreateFailedReplicas` makes more sense.

> In such case, the workflow does not finish at the end of `Reconcile` function.

We shouldn't rely on other functions to complete the workflow. It makes tracking down the lifetime difficult.

Reply:
Makes sense 👍